src/lib/utils.py

```diff
@@ -425,13 +433,17 @@ def dump_manifest(contents, manifest_path):
     """Writes back 'contents' to 'manifest_path'.
 
     For YAML, we make a best-effort attempt to preserve
-    formatting; for JSON, we use the canonical 4-space indentation."""
+    formatting; for JSON, we use the canonical 4-space indentation,
+    but add a trailing newline if originally present."""
     _, ext = os.path.splitext(manifest_path)
+    newline = _check_newline(manifest_path)
     with open(manifest_path, "w", encoding="utf-8") as fp:
```

Comment:
It would be a bit better if we could avoid opening the output file twice. Maybe open it in read-write mode here, and pass `fp` to `_check_newline`? Just make sure to `seek` the file to proper positions when using read-write mode.

Reply:
Sounds good, I'll change it and push a new commit.